django/db/models/query.py

```diff
@@ -1719,3 +1721,36 @@ def get_related_populators(klass_info, select, db):
         rel_cls = RelatedPopulator(rel_klass_info, select, db)
         iterators.append(rel_cls)
     return iterators
+
+
+class FilteredRelation(object):
+    def __init__(self, relation_name, alias, condition):
+        if not relation_name:
+            raise exceptions.FieldError("FilterRelation expects a non-empty relation_name")
+        if not alias:
+            raise exceptions.FieldError("FilterRelation expects a non-empty alias")
+        if relation_name == alias:
+            raise exceptions.FieldError("FilterRelation must be used with an alias %r"
+                                        " different from relation_name %r" % (alias, relation_name))
+        self.relation_name = relation_name
+        self.alias = alias
+        self.condition = condition
+        self.path = []
+
+    def __eq__(self, other):
+        if isinstance(other, self.__class__):
+            return (self.relation_name == other.relation_name and
+                    self.alias == other.alias and
+                    self.condition == other.condition)
+        return False
```

Comment:
one line

```py
return (
    isinstance(other, self.__class__) and
    self.relation_name == other.relation_name and
    self.alias == other.alias and
    self.condition == other.condition
)
```

`and` shortcuts, so it won't try get `other.relation_name` if it's not an instance of the same class 😉